Add operation.with_directives decorator

b-butler opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->

## Description
<!-- Describe your changes in detail -->
Adds a new decorator which add the appropriate directives along with registering the function as an operation `FlowProject.operation.with_directives`.

To accomplish this, we create a new class `OperationRegister`, that is specific to a `FlowProject` class like the preconditions and postconditions classes. We use the `__call__` magic method for the basic decorator.

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Addresses: #309

## Types of Changes
<!-- Please select all items that apply either now or after creating the pull request: -->
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
<!-- Please select all items that apply either now or after creating the pull request. -->
<!-- If you are unsure about any of these items, do not hesitate to ask! -->
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [x] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

bdice commented on tests/test_project.py:
```diff
@@ -708,6 +708,21 @@ def b(job):
             not callable(value) for value in submit_job_operation.directives.values()
         )
 
+    def test_operation_with_directies(self):
```
```suggestion
    def test_operation_with_directives(self):
```

bdice commented on flow/project.py:
```diff
@@ -1388,6 +1391,131 @@ def copy_from(cls, *other_funcs):
         return post
 
+    @staticmethod
+    def _setup_operation_register_object(parent_class):
```
This name is awkward -- it's long and the words don't tell me very much. I would prefer `_setup_operation_class`, to mirror `_setup_preconditions_class` and `_setup_postconditions_class`.

bdice commented on flow/project.py:
```diff
@@ -1388,6 +1391,131 @@ def copy_from(cls, *other_funcs):
+                return func
+
+            @classmethod
+            def with_directives(cls, directives, name=None):
```
I would prefer 
```suggestion
            def with_directives(cls, name, **directives):
```
as originally proposed in #309.

b-butler replied:
We mention keeping the same API as in `FlowGroup.with_directives` which uses a dictionary.

bdice commented on doc/api.rst:
```diff
@@ -58,8 +59,11 @@ The FlowProject
-    :exclude-members: pre,post
+    :exclude-members: pre,post,operation
 
+.. automethod:: flow.FlowProject.operation(func, name=None)
```
Do you have to specify the arguments, or can sphinx detect them automatically?

b-butler replied:
Have to

bdice reviewed (commented):
@b-butler I was curious what this PR did, so I gave it a very quick read-through. I recognize it's early stage, but thought that some design feedback might be helpful.

b-butler commented:
Here is the documentation PR https://github.com/glotzerlab/signac-docs/pull/128

I also added tests for lines of code that code coverage was failing me for not testing even though I did not write those lines of code, in doing so I fixed a small bug.

bdice reviewed (changes requested):
We should also modify the test files like `define_test_project.py` so that they use a mixture of the old and new syntax, at least until we deprecate/remove `@directives`.

bdice commented on flow/project.py:
```diff
@@ -1388,6 +1389,125 @@ def copy_from(cls, *other_funcs):
+                    if i > 0 and parameter_value.default is inspect.Parameter.empty:
+                        raise ValueError(
+                            "Only the first argument in an operation argument may not have "
+                            f"a default value! ({name})"
```
Is this a potential issue for aggregation? The user could specify:

```python
@FlowProject.operation
@aggregator.groupsof(2)
def my_operation(job1, job2):
    print(job1, job2)
```
and I'm also unsure of how this handles `*jobs`.

b-butler replied:
This is also a current check (just copied and pasted code), so I don't think this PR should fix that.

bdice commented on flow/project.py:
```diff
@@ -4184,6 +4231,12 @@ def foo(job):
+            name == operation_name for operation_name, _ in cls._OPERATION_FUNCTIONS
+        ):
+            raise ValueError(
+                f"Cannot create user group with name of existing operation {name}"
```
"User group" makes this sound like a UNIX authentication problem...
```suggestion
                f"Cannot create a group with the same name as the existing operation {name}"
```

bdice commented on tests/define_test_project.py:
```diff
@@ -53,20 +56,19 @@ def _need_to_fork(job):
 @_TestProject.post.false("test3_false")
 @_TestProject.post.not_(lambda job: job.doc.test3_false)
-@flow.directives(ngpu=1, omp_num_threads=1)
+@flow.directives()
```
Looks like this was left by mistake.
```suggestion
```

bdice commented on tests/test_project.py:
```diff
@@ -708,6 +734,38 @@ def b(job):
+        class A(FlowProject):
+            pass
+
+        # add deprecation warning context manager in v0.15
```
```suggestion
        # TODO: Add deprecation warning context manager in v0.15
```

bdice reviewed (approved):
I have a few smaller requests / questions. I approve after those comments are addressed. @lyrivera @kidrahahjo You're both assigned to this PR, could one of you review? Thanks!

bdice commented:
@b-butler Can you resolve merge conflicts?

b-butler commented:
@bdice Done.

kidrahahjo reviewed (approved):
I really like this implementation. These changes look good to me. @b-butler @bdice Is there a need to add documentation for this in signac-docs?

b-butler commented:
@kidrahahjo there is a docs PR https://github.com/glotzerlab/signac-docs/pull/128
